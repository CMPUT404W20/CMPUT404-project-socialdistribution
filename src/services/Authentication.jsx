import axios from "axios";

export const loginUser = (username, password) => {
    return axios.post("/auth/login/", {
        username: username,
        password: password,
    });
}

export const registerUser = (username, password) => {
    return axios.post("/auth/registration/", {
        username: username,
        password1: password,
        password2: password,
    });
}
