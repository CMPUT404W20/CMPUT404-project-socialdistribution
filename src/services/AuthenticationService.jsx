/* eslint-disable object-shorthand */
/* eslint-disable arrow-body-style */
import axios from "axios";
import Cookies from "js-cookie";

export const loginUser = (username, password) => {
  return axios.post("/auth/login/", {
    username,
    password,
  });
};

export const registerUser = (username, password) => {
  return axios.post("/auth/registration/", {
    username,
    password1: password,
    password2: password,
  });
};

export const getCurrentUser = () => {
  return axios.get("/author/current/");
};

export const logoutUser = () => {
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };

  return axios.post("/auth/logout/", null, { headers: headers });
};
