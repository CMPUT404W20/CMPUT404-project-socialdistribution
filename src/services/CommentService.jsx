/* eslint-disable no-alert */
/* eslint-disable no-console */
/* eslint-disable arrow-body-style */
import axios from "axios";

export const createComment = (newComment, postId) => {
  return axios.post(`posts/${postId}/comments/`, {
    comment: newComment,
  });
};

export const getComment = () => {
  return null;
};
