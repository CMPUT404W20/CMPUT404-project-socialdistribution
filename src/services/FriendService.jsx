/* eslint-disable arrow-body-style */
import axios from "axios";
import Cookies from "js-cookie";

export const getAuthorFriends = (authorId) => {
  return axios.get(`/author/${authorId}`).then((response) => {
    if (response.status === 200) {
      if (response.data && response.data.friends) {
        return response.data.friends;
      }
      return {};
    }

    throw new Error("Unable to retrieve friends");
  });
};

export const getAuthorFriendRequests = () => {
  return axios.get("/friendrequest").then((response) => {
    if (response.status === 200) {
      if (response.data) {
        return response.data;
      }
      return {};
    }

    throw new Error("Unable to retrieve friends");
  });
};

export const checkFriendStatus = (authorId1, authorId2) => {
  return axios.get(`/author/${authorId1}/friends/${authorId2}`).then((response) => {
    if (response.status === 200) {
      return response.data.friends;
    }
    return false;
  });
};

export const checkFollowingStatus = (authorId) => {
  return axios.get(`/following/${authorId}`).then((response) => {
    if (response.status === 200) {
      return response.data.following;
    }
    return false;
  });
};

export const unFriend = (friend) => {
  const payload = {
    query: "unfriend",
    friend,
  };
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };
  // eslint-disable-next-line object-shorthand
  return axios.post("/friend/unfriend/", payload, { headers }).then((response) => {
    if (response.status === 200) {
      return response.data.success;
    }
    throw new Error("Unable to send friend request");
  });
};

export const sendFriendRequest = (author, friend) => {
  const payload = {
    query: "friendrequest",
    author,
    friend,
  };
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };
  // eslint-disable-next-line object-shorthand
  return axios.post("/friendrequest", payload, { headers }).then((response) => {
    if (response.status === 201) {
      return response.data.success;
    }
    throw new Error("Unable to send friend request");
  });
};

export const rejectFriendRequest = (author, friend) => {
  const payload = {
    query: "friendrequest",
    author,
    friend,
  };
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };
  // eslint-disable-next-line object-shorthand
  return axios.post("/friendrequest/reject/", payload, { headers }).then((response) => {
    if (response.status === 204) {
      return true;
    }
    throw new Error("Unable to send friend request");
  });
};

export const searchAuthors = (displayName) => {
  return axios.get(`/author/search/${displayName}`).then((response) => {
    if (response.status === 200) {
      return response.data.authors;
    }
    return {};
  });
};

export const getAllAuthors = () => {
  return axios.get("/author/all/").then((response) => {
    if (response.status === 200) {
      return response.data;
    }

    return [];
  });
};

// for eslint
export const createFriend = () => {
  return null;
};
