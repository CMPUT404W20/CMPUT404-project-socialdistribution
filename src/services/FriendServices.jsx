/* eslint-disable arrow-body-style */
import axios from "axios";

export const getAuthorFriends = (authorId) => {
  return axios.get(`/author/${authorId}`).then((response) => {
    if (response.status === 200) {
      if (response.data && response.data.friends) {
        return response.data.friends;
      }
      return {};
    }

    throw new Error("Unable to retrieve friends");
  });
};

export const createFriend = () => {
  return null;
};
