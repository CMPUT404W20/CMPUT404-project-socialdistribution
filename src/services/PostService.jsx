/* eslint-disable object-shorthand */
/* eslint-disable arrow-body-style */
import axios from "axios";
import Cookies from "js-cookie";

export const getPosts = () => {
  return axios.get("/author/posts/").then((response) => {
    if (response.status === 200) {
      if (response.data && response.data.posts) {
        return response.data;
      }
      return {};
    }

    throw new Error("Unable to retrieve posts");
  });
};

export const getUserPosts = (fullUserId) => {
  return axios.get(`/author/${fullUserId}/posts/`).then((response) => {
    if (response.status === 200) {
      if (response.data && response.data.posts) {
        return response.data;
      }
      return {};
    }

    throw new Error("Unable to retrieve posts");
  });
};

export const createUserPosts = (postData) => {
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };

  // eslint-disable-next-line object-shorthand
  return axios.post("/author/posts/", postData, { headers: headers }).then((response) => {
    if (response.status === 201) {
      return response.data.success;
    }

    throw new Error("Unable to create post");
  });
};

export const deleteUserPosts = (postId) => {
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };

  return axios.delete(`/posts/${postId}/`, { headers: headers }).then((response) => {
    if (response.status === 204) {
      return {};
    }

    throw new Error("Unable to delete posts");
  });
};

export const updateUserPosts = (post) => {
  const csrf = Cookies.get("csrftoken");
  const headers = {
    "X-CSRFToken": csrf,
  };

  return axios.put(`/posts/${post.id}/`, post, { headers: headers }).then((response) => {
    if (response.status === 200) {
      return {};
    }

    throw new Error("Unable to update posts");
  });
};

