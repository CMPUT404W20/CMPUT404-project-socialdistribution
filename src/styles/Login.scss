%button {
    height: 47px;
    outline: 0;
    border-radius: 2px;
    box-shadow: 0px 0px 10px rgba(121, 121, 121, 0.25);
    font-weight: 500;
    cursor: pointer;
}

$accent-color: #4B84F2;

.login {
    background-color: grey;
    margin: 0;

    .login-description {
        background-color: #bbbbbb;
        height: 100vh;
    }

    .login-form {
        background-color: #ffffff;
        height: 100vh;
        padding: 0 50px 0;

        .login-message {
            font-size: 24px;
            font-weight: 500;
            letter-spacing: -0.05em;
            margin-bottom: 46px;
            margin-top: 30%;
        }

        .login-field {
            width: 100%;
            outline: 0;
            margin-top: 10px;
            border: 1px #EBEBEB solid;
            border-radius: 2px;
            font-size: 16px;
            padding: 15px 20px;

            &:focus {
                border: 1px #E0E0E0 solid;
            }
        }

        .login-button {
            @extend %button;
            width: 100%;
            height: 47px;
            background: $accent-color;
            color: white;
            border: 0px;
            margin-top: 76px;
        }
    }

    .login-secondary-options {
        margin-top: 50px;
        font-size: 14px;
        .login-secondary-options-text {
            line-height: 47px;
        }
    
        .login-secondary-button {
            @extend %button;
            width: 100px;
            float: right;
            border: 1px $accent-color solid;
            color: $accent-color;
        }
    }
}