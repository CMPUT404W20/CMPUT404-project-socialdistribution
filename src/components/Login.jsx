import React, { Component } from "react";
import "../styles/Login.scss";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import * as auth from "../services/Authentication";
import cover from "../images/cover.svg";

class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      signup: true,
      error: "",
      username: "",
      password: "",
      passwordReentry: "",
    };
  }

  handleInputChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  handleFormSubmit = (event) => {
    event.preventDefault();

    if (!this.validateForm()) { return; }

    const { signup, username, password } = this.state;

    if (signup) {
      auth.registerUser(username, password).then((response) => {
        if (response.status === 201) {
          this.setState({
            signup: false,
            username: "",
            password: "",
          });
        }
      }).catch((err) => {
        let error = err.response.data;
        error = error.username || error.password1 || error.password2;
        this.setState({
          error: error ? error[0] : "Something went wrong",
        });
      });
    } else {
      auth.loginUser(username, password).then((response) => {
        if (response.status === 200) {
          // TODO: success: redirect to homepage
          // eslint-disable-next-line no-alert
          alert("Success");
        }
      }).catch(() => {
        this.setState({
          error: "Username or Password are incorrect",
        });
      });
    }
  }

  toggleMethod = () => {
    // toggle between sign-up and sign-in screens
    const { signup } = this.state;

    this.setState({
      signup: !signup,
      error: "",
    });
  }

  validateForm = () => {
    const {
      signup, username, password, passwordReentry,
    } = this.state;

    let error = "";
    if (!username) {
      error = "Please fill in the username";
    } else if (!password) {
      error = "Please fill in the password";
    } else if (signup) {
      if (!passwordReentry) {
        error = "Please confirm your password";
      } else if (password !== passwordReentry) {
        error = "Passwords don't match";
      } else if (password.length < 8) {
        // this is enforced by Django Auth so we need to check here
        error = "Password must be atleast 8 characters long";
      }
    }

    this.setState({
      error,
    });

    return !error;
  }

  renderForm() {
    const {
      signup, username, password, passwordReentry, error,
    } = this.state;

    let formMessage = "Let's Get Started!";
    let primaryButtonText = "Sign up";
    if (!signup) {
      formMessage = "Welcome Back!";
      primaryButtonText = "Sign in";
    }

    return (
      <form onSubmit={this.handleFormSubmit}>
        <div className="login-message">
          {formMessage}
        </div>

        <input
          type="text"
          name="username"
          className="login-field"
          placeholder="Username"
          onChange={this.handleInputChange}
          value={username}
        />

        <input
          type="password"
          name="password"
          className="login-field"
          placeholder="Password"
          onChange={this.handleInputChange}
          value={password}
        />

        {
          signup ? (
            <input
              type="password"
              name="passwordReentry"
              className="login-field"
              placeholder="Re-enter Password"
              onChange={this.handleInputChange}
              value={passwordReentry}
            />
          )
            : null
        }

        <span className="login-error-message">{error}</span>
        <button className="login-button" type="submit">{primaryButtonText}</button>
      </form>
    );
  }

  renderSecondaryOption() {
    const { signup } = this.state;

    let secondaryOptionText = "Already have account?";
    let secondaryButtonText = "Sign in";
    if (!signup) {
      secondaryOptionText = "Don't have an account?";
      secondaryButtonText = "Sign up";
    }

    return (
      <div className="login-secondary-options">
        <span className="login-secondary-options-text">{secondaryOptionText}</span>
        <button className="login-secondary-button" type="submit" onClick={this.toggleMethod}>{secondaryButtonText}</button>
      </div>
    );
  }

  render() {
    return (
      <Container fluid className="login">
        <Row>
          <Col md={7} className="login-description">
            <div className="login-heading">
              App Name
            </div>
            <div className="cover-image-wrapper">
              <img className="cover-image" src={cover} width="70%" alt="cover" />
            </div>
          </Col>
          <Col md={5} className="login-form-wrapper">
            {this.renderForm()}
            {this.renderSecondaryOption()}
          </Col>
        </Row>
      </Container>
    );
  }
}

export default Login;
