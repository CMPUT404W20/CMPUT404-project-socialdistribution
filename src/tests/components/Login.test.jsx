import React from "react";
import { shallow, mount } from "enzyme";
import Login from "../../components/Login";
import * as auth from "../../services/Authentication";

describe("Login Page", () => {
  it("should render correctly", () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });

  it("should switch to sign in on secondary button click", () => {
    const component = shallow(<Login />);
    const signinButton = component.find(".login-secondary-button");

    signinButton.simulate("click");

    expect(component).toMatchSnapshot();
  });

  it("should show error if fields are incorrect", () => {
    const wrapper = mount(<Login />);
    const signupButton = wrapper.find("form>.login-button");

    wrapper.setState({
      password: "password",
    });
    signupButton.simulate("submit");
    expect(wrapper.state("error")).toBe("Please fill in the username");

    wrapper.setState({
      username: "username",
      password: "password",
    });
    signupButton.simulate("submit");
    expect(wrapper.state("error")).toBe("Please confirm your password");

    wrapper.setState({
      username: "username",
      password: "password",
      passwordReentry: "mismatchedPassword",
    });
    signupButton.simulate("submit");
    expect(wrapper.state("error")).toBe("Passwords don't match");
  });

  it("should not send a request if there is an error", () => {
    const wrapper = mount(<Login />);
    const registerUserSpy = jest.spyOn(auth, "registerUser");

    wrapper.setState({
      error: "some error",
    });

    const signupButton = wrapper.find("form>.login-button");
    signupButton.simulate("submit");
    expect(registerUserSpy).toHaveBeenCalledTimes(0);

    wrapper.setState({
      signup: false, // switch to login mode
    });

    const signinButton = wrapper.find("form>.login-button");
    signinButton.simulate("submit");
    expect(registerUserSpy).toHaveBeenCalledTimes(0);
  });

  it("should send registration request if sign up fields are valid", () => {
    const wrapper = mount(<Login />);
    const signupButton = wrapper.find("form>.login-button");

    const registerUserSpy = jest.spyOn(auth, "registerUser");
    wrapper.setState({
      username: "username",
      password: "password",
      passwordReentry: "password",
    });

    signupButton.simulate("submit");
    expect(wrapper.state("error")).toBe("");
    expect(registerUserSpy).toHaveBeenCalledTimes(1);
  });

  it("should send login request if login fields are valid", () => {
    const wrapper = mount(<Login />);
    wrapper.setState({
      signup: false, // switch to login mode
    });
    const signinButton = wrapper.find("form>.login-button");

    const loginUserSpy = jest.spyOn(auth, "loginUser");
    wrapper.setState({
      username: "username",
      password: "password",
    });

    signinButton.simulate("submit");
    expect(wrapper.state("error")).toBe("");
    expect(loginUserSpy).toHaveBeenCalledTimes(1);
  });
});
