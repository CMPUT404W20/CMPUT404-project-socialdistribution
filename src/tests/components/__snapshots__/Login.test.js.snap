// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render correctly 1`] = `
<Container
  className="login"
  fluid={true}
>
  <ForwardRef
    noGutters={false}
  >
    <Col
      className="login-description"
      md={7}
    >
      <div
        className="login-heading"
      >
        App Name
      </div>
      <div
        className="cover-image-wrapper"
      >
        <img
          className="cover-image"
          src="cover.svg"
          width="70%"
        />
      </div>
    </Col>
    <Col
      className="login-form-wrapper"
      md={5}
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="login-message"
        >
          Let's Get Started!
        </div>
        <input
          autoFocus={true}
          className="login-field"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
        <input
          className="login-field"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <input
          className="login-field"
          name="passwordReentry"
          onChange={[Function]}
          placeholder="Re-enter Password"
          type="password"
          value=""
        />
        <span
          className="login-error-message"
        />
        <button
          className="login-button"
          type="submit"
        >
          Sign up
        </button>
      </form>
      <div
        className="login-secondary-options"
      >
        <span
          className="login-secondary-options-text"
        >
          Already have account?
        </span>
        <button
          className="login-secondary-button"
          onClick={[Function]}
          type="submit"
        >
          Sign in
        </button>
      </div>
    </Col>
  </ForwardRef>
</Container>
`;

exports[`Login Page should switch to sign in on secondary button click 1`] = `
<Container
  className="login"
  fluid={true}
>
  <ForwardRef
    noGutters={false}
  >
    <Col
      className="login-description"
      md={7}
    >
      <div
        className="login-heading"
      >
        App Name
      </div>
      <div
        className="cover-image-wrapper"
      >
        <img
          className="cover-image"
          src="cover.svg"
          width="70%"
        />
      </div>
    </Col>
    <Col
      className="login-form-wrapper"
      md={5}
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="login-message"
        >
          Welcome Back!
        </div>
        <input
          autoFocus={true}
          className="login-field"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
        <input
          className="login-field"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <span
          className="login-error-message"
        />
        <button
          className="login-button"
          type="submit"
        >
          Sign in
        </button>
      </form>
      <div
        className="login-secondary-options"
      >
        <span
          className="login-secondary-options-text"
        >
          Don't have an account?
        </span>
        <button
          className="login-secondary-button"
          onClick={[Function]}
          type="submit"
        >
          Sign up
        </button>
      </div>
    </Col>
  </ForwardRef>
</Container>
`;
